// @beenotung/tslib/gen-json.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.genNumber = genNumber;
exports.genString = genString;
exports.genNull = genNull;
exports.getBoolean = getBoolean;
exports.genJsonPrimitive = genJsonPrimitive;
exports.genJsonArray = genJsonArray;
exports.genJsonObject = genJsonObject;
exports.genJsonValue = genJsonValue;
const random_1 = require("./random");
function genNumber() {
    return random_1.Random.element([random_1.Random.nextInt, random_1.Random.nextFloat])();
}
function genString(length = 8) {
    return random_1.Random.nextString(length, random_1.visibleLetters);
}
function genNull() {
    return null;
}
function getBoolean() {
    return random_1.Random.nextBool();
}
function genJsonPrimitive(length = 8) {
    return random_1.Random.element([
        genString,
        genNumber,
        genNull,
        getBoolean,
    ])(length);
}
function genJsonArray(length = 8) {
    const xs = new Array(length);
    for (let i = 0; i < length; i++) {
        xs[i] = genJsonValue(length - 1);
    }
    return xs;
}
function genJsonObject(length = 8) {
    const o = {};
    for (let i = 0; i < length; i++) {
        o[genString(length)] = genJsonValue(length - 1);
    }
    return o;
}
function genJsonValue(length = 8) {
    return random_1.Random.element([
        genJsonPrimitive,
        genJsonArray,
        genJsonObject,
    ])(length);
}
